File: stripe/src/test/java/com/stripe/android/view/ExpiryDateEditTextTest.java
package com.stripe.android.view;

import com.stripe.android.BuildConfig;
import com.stripe.android.testharness.ViewTestUtils;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ActivityController;
import org.robolectric.annotation.Config;

import java.util.Calendar;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;

/**
 * Test class for {@link ExpiryDateEditText}.
 */
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 25)
public class ExpiryDateEditTextTest {

    @Mock ExpiryDateEditText.ExpiryDateEditListener mExpiryDateEditListener;
    private ExpiryDateEditText mExpiryDateEditText;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        ActivityController<CardInputTestActivity> activityController =
                Robolectric.buildActivity(CardInputTestActivity.class).create().start();

        mExpiryDateEditText = activityController.get().getExpiryDateEditText();
        mExpiryDateEditText.setText("");
        mExpiryDateEditText.setExpiryDateEditListener(mExpiryDateEditListener);
    }

    @Test
    public void afterInputTwoDigits_textContainsSlash() {
        mExpiryDateEditText.append("1");
        mExpiryDateEditText.append("2");

        String text = mExpiryDateEditText.getText().toString();
        assertEquals("12/", text);
    }

    @Test
    public void inputSingleDigit_whenDigitIsTooLargeForMonth_prependsZero() {
        mExpiryDateEditText.append("4");
        assertEquals("04/", mExpiryDateEditText.getText().toString());
        assertEquals(3, mExpiryDateEditText.getSelectionStart());
    }

    @Test
    public void inputSingleDigit_whenDigitIsZeroOrOne_doesNotPrependZero() {
        mExpiryDateEditText.append("1");
        assertEquals("1", mExpiryDateEditText.getText().toString());
        assertEquals(1, mExpiryDateEditText.getSelectionStart());
    }

    @Test
    public void inputSingleDigit_whenAtFirstCharacterButTextNotEmpty_doesNotPrependZero() {
        mExpiryDateEditText.append("1");
        // This case can occur when the user moves the cursor behind the already-entered text.
        mExpiryDateEditText.setSelection(0);
        mExpiryDateEditText.getEditableText().replace(0, 0, "3", 0, 1);

        assertEquals("31", mExpiryDateEditText.getText().toString());
        assertEquals(1, mExpiryDateEditText.getSelectionStart());
    }

    @Test
    public void inputMultipleValidDigits_whenEmpty_doesNotPrependZeroOrShowErrorState() {
        mExpiryDateEditText.append("11");

        String text = mExpiryDateEditText.getText().toString();
        assertEquals("11/", text);
        assertFalse(mExpiryDateEditText.getShouldShowError());
        assertEquals(3, mExpiryDateEditText.getSelectionStart());
    }

    @Test
    public void afterInputThreeDigits_whenDeletingOne_textDoesNotContainSlash() {
        mExpiryDateEditText.append("1");
        mExpiryDateEditText.append("2");
        mExpiryDateEditText.append("3");
        ViewTestUtils.sendDeleteKeyEvent(mExpiryDateEditText);
        String text = mExpiryDateEditText.getText().toString();
        assertEquals("12", text);
    }

    @Test
    public void afterAddingFinalDigit_whenGoingFromInvalidToValid_callsListener() {
        if (Calendar.getInstance().get(Calendar.YEAR) > 2059) {
            fail("Update the code with a date that is still valid. Also, hello from the past.");
        }

        mExpiryDateEditText.append("1");
        mExpiryDateEditText.append("2");
        mExpiryDateEditText.append("5");
        verifyZeroInteractions(mExpiryDateEditListener);

        mExpiryDateEditText.append("9");
        assertTrue(mExpiryDateEditText.isDateValid());
        verify(mExpiryDateEditListener, times(1)).onExpiryDateComplete();
    }

    @Test
    public void afterAddingFinalDigit_whenDeletingItem_revertsToInvalidState() {
        if (Calendar.getInstance().get(Calendar.YEAR) > 2059) {
            fail("Update the code with a date that is still valid. Also, hello from the past.");
        }

        mExpiryDateEditText.append("12");
        mExpiryDateEditText.append("59");
        assertTrue(mExpiryDateEditText.isDateValid());

        ViewTestUtils.sendDeleteKeyEvent(mExpiryDateEditText);
        verify(mExpiryDateEditListener, times(1)).onExpiryDateComplete();
        assertFalse(mExpiryDateEditText.isDateValid());
    }

    @Test
    public void updateSelectionIndex_whenMovingAcrossTheGap_movesToEnd() {
        assertEquals(3, mExpiryDateEditText.updateSelectionIndex(3, 1, 1, 5));
    }

    @Test
    public void updateSelectionIndex_atStart_onlyMovesForwardByOne() {
        assertEquals(1, mExpiryDateEditText.updateSelectionIndex(1, 0, 1, 5));
    }

    @Test
    public void updateSelectionIndex_whenDeletingAcrossTheGap_staysAtEnd() {
        assertEquals(2, mExpiryDateEditText.updateSelectionIndex(2, 4, 0, 5));
    }

    @Test
    public void updateSelectionIndex_whenInputVeryLong_respectMaxInputLength() {
        assertEquals(5, mExpiryDateEditText.updateSelectionIndex(6, 4, 2, 5));
    }

    @Test
    public void inputZero_whenEmpty_doesNotShowErrorState() {
        mExpiryDateEditText.append("0");

        assertFalse(mExpiryDateEditText.getShouldShowError());
        assertEquals("0", mExpiryDateEditText.getText().toString());
    }

    @Test
    public void inputOne_whenEmpty_doesNotShowErrorState() {
        mExpiryDateEditText.append("1");

        assertFalse(mExpiryDateEditText.getShouldShowError());
        assertEquals("1", mExpiryDateEditText.getText().toString());
    }

    @Test
    public void inputTwoDigitMonth_whenInvalid_showsErrorAndDoesNotAddSlash() {
        mExpiryDateEditText.append("14");

        assertTrue(mExpiryDateEditText.getShouldShowError());
        assertEquals("14", mExpiryDateEditText.getText().toString());
    }

    @Test
    public void inputThreeDigits_whenInvalid_showsErrorAndDoesAddSlash() {
        mExpiryDateEditText.append("143");

        assertTrue(mExpiryDateEditText.getShouldShowError());
        assertEquals("14/3", mExpiryDateEditText.getText().toString());
    }

    @Test
    public void delete_whenAcrossSeparator_alwaysDeletesNumber() {
        mExpiryDateEditText.append("12");
        assertEquals("12/", mExpiryDateEditText.getText().toString());

        ViewTestUtils.sendDeleteKeyEvent(mExpiryDateEditText);
        assertEquals("1", mExpiryDateEditText.getText().toString());
    }

    @Test
    public void inputCompleteDate_whenInPast_showsInvalid() {
        // This test will be invalid if run between 2080 and 2112. Please update the code.
        assertTrue(Calendar.getInstance().get(Calendar.YEAR) < 2080);
        // Date translates to December, 2012 UNTIL the 2080, at which point it translates to
        // December, 2112. Also, this simulates a PASTE action.
        mExpiryDateEditText.append("1212");

        assertTrue(mExpiryDateEditText.getShouldShowError());
        verifyZeroInteractions(mExpiryDateEditListener);
    }

    @Test
    public void inputCompleteDateInPast_thenDelete_showsValid() {
        // This test will be invalid if run between 2080 and 2112. Please update the code.
        assertTrue(Calendar.getInstance().get(Calendar.YEAR) < 2080);
        // Date translates to December, 2012 UNTIL the 2080, at which point it translates to
        // December, 2112.
        mExpiryDateEditText.append("12/12");
        assertTrue(mExpiryDateEditText.getShouldShowError());

        ViewTestUtils.sendDeleteKeyEvent(mExpiryDateEditText);
        assertEquals("12/1", mExpiryDateEditText.getText().toString());
        assertFalse(mExpiryDateEditText.getShouldShowError());

        // The date is no longer "in error", but it still shouldn't have triggered the listener.
        verifyZeroInteractions(mExpiryDateEditListener);
    }

    @Test
    public void getValidDateFields_whenDataIsValid_returnsExpectedValues() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertTrue(Calendar.getInstance().get(Calendar.YEAR) < 2050);

        mExpiryDateEditText.append("12");
        mExpiryDateEditText.append("50");

        int [] retrievedDate = mExpiryDateEditText.getValidDateFields();
        assertNotNull(retrievedDate);
        assertEquals(12, retrievedDate[0]);
        assertEquals(2050, retrievedDate[1]);
    }

    @Test
    public void getValidDateFields_whenDateIsValidFormatButExpired_returnsNull() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertTrue(Calendar.getInstance().get(Calendar.YEAR) < 2080);

        mExpiryDateEditText.append("12");
        mExpiryDateEditText.append("12");
        // 12/12 is an invalid date until 2080, at which point it will be interpreted as 12/2112

        assertNull(mExpiryDateEditText.getValidDateFields());
    }

    @Test
    public void getValidDateFields_whenDateIsIncomplete_returnsNull() {
        mExpiryDateEditText.append("4");
        assertNull(mExpiryDateEditText.getValidDateFields());
    }

    @Test
    public void getValidDateFields_whenDateIsValidAndThenChangedToInvalid_returnsNull() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertTrue(Calendar.getInstance().get(Calendar.YEAR) < 2050);

        mExpiryDateEditText.append("12");
        mExpiryDateEditText.append("50");
        ViewTestUtils.sendDeleteKeyEvent(mExpiryDateEditText);

        assertNull(mExpiryDateEditText.getValidDateFields());
    }
}

File: stripe/src/test/java/com/stripe/android/view/ExpiryDateEditTextTest.kt
package com.stripe.android.view

import androidx.appcompat.view.ContextThemeWrapper
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import com.stripe.android.R
import com.stripe.android.model.ExpirationDate
import com.stripe.android.testharness.ViewTestUtils
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.Calendar
import kotlin.test.Test

/**
 * Test class for [ExpiryDateEditText].
 */
@RunWith(RobolectricTestRunner::class)
class ExpiryDateEditTextTest {
    private val expiryDateEditText = ExpiryDateEditText(
        ContextThemeWrapper(
            ApplicationProvider.getApplicationContext(),
            R.style.StripeDefaultTheme
        )
    )

    @Test
    fun afterInputTwoDigits_textContainsSlash() {
        expiryDateEditText.append("1")
        expiryDateEditText.append("2")

        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("12/")
    }

    @Test
    fun inputSingleDigit_whenDigitIsTooLargeForMonth_prependsZero() {
        expiryDateEditText.append("4")
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("04/")
        assertThat(expiryDateEditText.selectionStart)
            .isEqualTo(3)
    }

    @Test
    fun inputSingleDigit_whenDigitIsZeroOrOne_doesNotPrependZero() {
        expiryDateEditText.append("1")
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("1")
        assertThat(expiryDateEditText.selectionStart)
            .isEqualTo(1)
    }

    @Test
    fun inputSingleDigit_whenAtFirstCharacterButTextNotEmpty_doesNotPrependZero() {
        expiryDateEditText.append("1")
        // This case can occur when the user moves the cursor behind the already-entered text.
        expiryDateEditText.setSelection(0)
        expiryDateEditText.editableText.replace(0, 0, "3", 0, 1)

        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("31")
        assertThat(expiryDateEditText.selectionStart)
            .isEqualTo(1)
    }

    @Test
    fun inputMultipleValidDigits_whenEmpty_doesNotPrependZeroOrShowErrorState() {
        expiryDateEditText.append("11")

        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("11/")
        assertThat(expiryDateEditText.shouldShowError)
            .isFalse()
        assertThat(expiryDateEditText.selectionStart)
            .isEqualTo(3)
    }

    @Test
    fun afterInputThreeDigits_whenDeletingOne_textDoesContainSlash() {
        expiryDateEditText.append("1")
        expiryDateEditText.append("2")
        expiryDateEditText.append("3")

        ViewTestUtils.sendDeleteKeyEvent(expiryDateEditText)
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("12/")
    }

    @Test
    fun afterAddingFinalDigit_whenGoingFromInvalidToValid_callsListener() {
        var invocations = 0
        expiryDateEditText.completionCallback = {
            invocations++
        }

        assertThat(Calendar.getInstance().get(Calendar.YEAR) <= 2059)
            .isTrue()

        expiryDateEditText.append("1")
        expiryDateEditText.append("2")
        expiryDateEditText.append("5")
        assertThat(invocations)
            .isEqualTo(0)

        expiryDateEditText.append("9")
        assertThat(expiryDateEditText.isDateValid)
            .isTrue()
        assertThat(invocations)
            .isEqualTo(1)
    }

    @Test
    fun afterAddingFinalDigit_whenDeletingItem_revertsToInvalidState() {
        var invocations = 0
        expiryDateEditText.completionCallback = {
            invocations++
        }

        assertThat(Calendar.getInstance().get(Calendar.YEAR) <= 2059)
            .isTrue()

        expiryDateEditText.append("12")
        expiryDateEditText.append("59")
        assertThat(expiryDateEditText.isDateValid)
            .isTrue()

        ViewTestUtils.sendDeleteKeyEvent(expiryDateEditText)
        assertThat(invocations)
            .isEqualTo(1)
        assertThat(expiryDateEditText.isDateValid)
            .isFalse()
    }

    @Test
    fun updateSelectionIndex_whenMovingAcrossTheGap_movesToEnd() {
        assertThat(
            expiryDateEditText.updateSelectionIndex(3, 1, 1, 5)
        ).isEqualTo(3)
    }

    @Test
    fun updateSelectionIndex_atStart_onlyMovesForwardByOne() {
        assertThat(
            expiryDateEditText.updateSelectionIndex(1, 0, 1, 5)
        ).isEqualTo(1)
    }

    @Test
    fun updateSelectionIndex_whenDeletingAcrossTheGap_staysAtEnd() {
        assertThat(
            expiryDateEditText.updateSelectionIndex(2, 4, 0, 5)
        ).isEqualTo(2)
    }

    @Test
    fun updateSelectionIndex_whenInputVeryLong_respectMaxInputLength() {
        assertThat(
            expiryDateEditText.updateSelectionIndex(6, 4, 2, 5)
        ).isEqualTo(5)
    }

    @Test
    fun inputZero_whenEmpty_doesNotShowErrorState() {
        expiryDateEditText.append("0")

        assertThat(expiryDateEditText.shouldShowError)
            .isFalse()
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("0")
    }

    @Test
    fun inputOne_whenEmpty_doesNotShowErrorState() {
        expiryDateEditText.append("1")

        assertThat(expiryDateEditText.shouldShowError)
            .isFalse()
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("1")
    }

    @Test
    fun inputTwoDigitMonth_whenInvalid_showsErrorAndDoesNotAddSlash() {
        expiryDateEditText.append("14")

        assertThat(expiryDateEditText.shouldShowError)
            .isTrue()
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("14")
    }

    @Test
    fun inputThreeDigits_whenInvalid_showsErrorAndDoesAddSlash() {
        expiryDateEditText.append("143")

        assertThat(expiryDateEditText.shouldShowError)
            .isTrue()
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("14/3")
    }

    @Test
    fun delete_whenAcrossSeparator_deletesSeparator() {
        expiryDateEditText.append("12")
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("12/")

        ViewTestUtils.sendDeleteKeyEvent(expiryDateEditText)
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("12")
    }

    @Test
    fun inputCompleteDate_whenInPast_showsInvalid() {
        var invocations = 0
        expiryDateEditText.completionCallback = {
            invocations++
        }

        // This test will be invalid if run between 2080 and 2112. Please update the code.
        assertThat(Calendar.getInstance().get(Calendar.YEAR) < 2080)
            .isTrue()
        // Date translates to December, 2012 UNTIL the 2080, at which point it translates to
        // December, 2112. Also, this simulates a PASTE action.
        expiryDateEditText.append("1212")

        assertThat(expiryDateEditText.shouldShowError)
            .isTrue()
        assertThat(invocations)
            .isEqualTo(0)
    }

    @Test
    fun inputCompleteDateInPast_thenDelete_showsValid() {
        var invocations = 0
        expiryDateEditText.completionCallback = {
            invocations++
        }

        // This test will be invalid if run between 2080 and 2112. Please update the code.
        assertThat(Calendar.getInstance().get(Calendar.YEAR) < 2080)
            .isTrue()
        // Date translates to December, 2012 UNTIL the 2080, at which point it translates to
        // December, 2112.
        expiryDateEditText.append("12/12")
        assertThat(expiryDateEditText.shouldShowError)
            .isTrue()

        ViewTestUtils.sendDeleteKeyEvent(expiryDateEditText)
        assertThat(expiryDateEditText.text.toString())
            .isEqualTo("12/1")
        assertThat(expiryDateEditText.shouldShowError)
            .isFalse()

        // The date is no longer "in error", but it still shouldn't have triggered the listener.
        assertThat(invocations)
            .isEqualTo(0)
    }

    @Test
    fun validatedDate_whenDataIsValid_returnsExpectedValues() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertThat(Calendar.getInstance().get(Calendar.YEAR) < 2050)
            .isTrue()

        expiryDateEditText.append("12")
        expiryDateEditText.append("50")

        assertThat(
            expiryDateEditText.validatedDate
        ).isEqualTo(
            ExpirationDate.Validated(
                month = 12,
                year = 2050
            )
        )
    }

    @Test
    fun validatedDate_whenDateIsValidFormatButExpired_returnsNull() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertThat(Calendar.getInstance().get(Calendar.YEAR) < 2080)
            .isTrue()

        expiryDateEditText.append("12")
        expiryDateEditText.append("12")
        // 12/12 is an invalid date until 2080, at which point it will be interpreted as 12/2112

        assertThat(expiryDateEditText.validatedDate)
            .isNull()
    }

    @Test
    fun validatedDate_whenDateIsIncomplete_returnsNull() {
        expiryDateEditText.append("4")
        assertThat(expiryDateEditText.validatedDate)
            .isNull()
    }

    @Test
    fun validatedDate_whenDateIsValidAndThenChangedToInvalid_returnsNull() {
        // This test will be invalid if run after the year 2050. Please update the code.
        assertThat(Calendar.getInstance().get(Calendar.YEAR) < 2050)
            .isTrue()

        expiryDateEditText.append("12")
        expiryDateEditText.append("50")
        ViewTestUtils.sendDeleteKeyEvent(expiryDateEditText)

        assertThat(expiryDateEditText.validatedDate)
            .isNull()
    }
}
